Avoid over iterating when filtering

otaviolima opened the pull request:
Use array.first(where:) instead of array.filter().first
Use array.contains(where:) instead filtering and checking for filtered array size
Use array.reversed().first(where:) instead of array.filter().last

allenhumphreys commented on ConfCore/Session.swift:
```diff
@@ -67,7 +67,7 @@ public class Session: Object {
         guard let realm = self.realm else { return nil }
         guard !transcriptIdentifier.isEmpty else { return nil }
         
-        return realm.objects(Transcript.self).filter("identifier == %@", self.transcriptIdentifier).first
```
This `filter` method is a realm query not a filter loop. I'm pretty sure it gets turned into SQLLite query.

The first code block from [the docs](https://realm.io/docs/swift/latest/) includes an example `let theDog = realm.objects(Dog.self).filter("age == 1").first`

Though, from a Swift perspective, this is generally the correct move.

For an explanation of why [see this](https://realm.io/docs/swift/2.10.0/api/Classes/Results.html#/s:FC10RealmSwift7Results6filterFtSSGSaP___GS0_x_)
>`Results` are lazily evaluated the first time they are accessed; they only run queries when the result of the query is requested. This means that chaining several temporary `Results` to sort and filter your data does not perform any unnecessary work processing the intermediate state.

insidegui replied:
@otaviolima Did you perform any measurements to confirm that your changes actually make things faster?

@allenhumphreys Realm is not SQL-based, but from my understanding, when using predicate formats for filtering it is more performant. I might be wrong tho, maybe we can ask someone from Realm to clear things up regarding their filtering methods.

insidegui replied:
@otaviolima BTW, I'm only talking about the places where predicate formats got replaced by closures, most of the changes are clearly for the better ;)

allenhumphreys commented:
@insidegui haha, you're right about the tech, but I still believe in the general point I was making 😬

otaviolima commented:
@insidegui As @allenhumphreys pointed out the Realm code should remain as it it.

I tested and the performance drop is huge.

I removed all the changes that were related to Realm, rebased and pushed again.
